.github/actions/sam-build/action.yaml

```diff
@@ -0,0 +1,24 @@
+name: sam-build
+
+runs:
+  # This creates a composite action to be used as a step in a job
+  # https://docs.github.com/en/actions/creating-actions/creating-a-composite-action
+  using: "composite"
```

Comment:
I'm wondering why you chose to use a composite action instead of reusable workflow?

Reply:
I wanted to create two completely separate workflows for pre- and post-merge so that we only ever attempt to get a write-permission token in the post-merge workflow (a belt & suspenders approach). but I didn't want to duplicate the steps for the tests (or skip them during the post-merge workflow), so I created a composite action to share the steps between the workflows.

Reply:
could you not do the same thing with a reusable workflow?  i'm asking because i feel that reusable workflows are usually more flexible and conducive to more use cases than a composite action.

Reply:
not really, the steps in the composite action are making changes to the local filesystem, but each workflow runs with its own checkout of the repo. the goal is to be able to run `sam build` in the pre-merge workflow with a read-only token before calling `sam validate`, and also run it in the post-merge workflow before running `sam package` and uploading the result to AWS with a write token. if I move the build from the composite action to a reuseable workflow, I'm not sure how to share the output of `sam build` with either `sam validate` or `sam package`.

Reply:
what's the motivation behind sharing the output since `sam build` is a relatively quick step?

Reply:
> what's the motivation behind sharing the output 
I don't fully understand this question, either the commands need to be run against the same filesystem by running them as steps in the same job, or we need a way of copying the files from one job context to another.

one goal here is to create a single definition for the steps needed to setup sam (i.e. install sam, generate `requirements.txt`, and build the `.aws-sam/build/` directory that will be zipped) so that if we ever make changes to that process we do it in a single location. and another goal is to isolate write tokens to a post-merge workflow.

so I think a composite action is more appropriate because that allows us to define the steps once but run them separately in multiple jobs; but a reusable workflow would run the steps once and require uploading the generated files to a workflow artifact and download them in the next job, and that seems more difficult to do correctly (are the artifacts versioned or given unique names? or are they overwritten by the next workflow run? etc.). the composite action seems simpler to me.

Reply:
ahh, ok.  i think you are saying that a resuable workflow runs as a separate process outside the context of the main action which doesn't work well for this use case?

Reply:
yeah, exactly